### PR TITLE
Fix build on systems without ucontext.h in libc

### PR DESCRIPTION
**Pull request by pavelkumbrasev**
### Description 
This patch is based on the @ismaell contribution in [PR#900](https://github.com/oneapi-src/oneTBB/pull/900).
Enable thread based resumable tasks if ucontext is absent in libc.
The performance of thread based implementation is questionable but at least oneTBB can be built. 

Fixes # - _issue number(s) if exists_

- [x] - git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [] bug fix - _change that fixes an issue_
- [x] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [ ] added - _required for new features and some bug fixes_
- [x] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [x] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
@ismaell @brad0

### CODE REVIEW

**Comment by pavelkumbrasev** [comment]
@vossmjp Do you ok with such approach?

File: cmake/resumable_tasks.cmake [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include(CheckSymbolExists)
+
+if (CMAKE_SYSTEM_NAME MATCHES "^Linux")
```
Comment (isaevil) [inline cmake/resumable_tasks.cmake:17]:
```suggestion
if (UNIX)
```

File: cmake/resumable_tasks.cmake [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    check_symbol_exists("getcontext" "ucontext.h" HAVE_UCONTEXT)
+    if (HAVE_UCONTEXT)
```
Comment (isaevil) [inline cmake/resumable_tasks.cmake:19]: I would rename `HAVE_UCONTEXT` to `_tbb_have_ucontext` since it is not meant to be used globally.

**Review by sarathnandu — APPROVED** [review]
LGTM!
The CI failures seems to be non-related.

**Comment by sarathnandu** [comment]
https://github.com/oneapi-src/oneTBB/blob/master/include/oneapi/tbb/detail/_config.h#L459
IS this still correct?

**Comment by brad0** [comment]
Builds. Ran the test suite..

```
93% tests passed, 9 tests failed out of 135

Total Test time (real) = 424.06 sec

The following tests FAILED:
          4 - test_dynamic_link (Failed)
         17 - test_parallel_invoke (Subprocess aborted)
         23 - test_task_group (Subprocess aborted)
         68 - test_arena_constraints (Subprocess aborted)
        125 - test_malloc_regression (Subprocess aborted)
        127 - test_malloc_compliance (Subprocess aborted)
        128 - test_malloc_used_by_lib (Subprocess aborted)
        129 - test_malloc_lib_unload (Subprocess aborted)
        131 - test_malloc_whitebox (Subprocess aborted)
Errors while running CTest
```

Had to workaround two minor issues..

https://github.com/oneapi-src/oneTBB/issues/1015

**Comment by brad0** [comment]
You should remove the APPLE bits from src/tbb/co_context.h and have the thread implementation for macOS as well.

**Comment by brad0** [comment]
ping.

**Comment by brad0** [comment]
ping.

**Comment by pavelkumbrasev** [comment]
@brad0 Just to clarify. 
The real issue with resumable task for you, is that you can not build oneTBB because of ucontext absence? 
I curious because while thread-based implementation will definitely solve build problem it will contain comparably worse (vs ucontext based) performance for resumable tasks.

**Comment by brad0** [comment]
> @brad0 Just to clarify. The real issue with resumable task for you, is that you can not build oneTBB because of ucontext absence? I curious because while thread-based implementation will definitely solve build problem it will contain comparably worse (vs ucontext based) performance for resumable tasks.

Correct. No ucontext API at all. Modern POSIX compliant OS's have no requirement to implement said API as it is deprecated.

**Comment by brad0** [comment]
This isn't the first time dealing with something like this. QEMU is another example with their use of ucontext for co-routines.
The threaded backend is used on macOS and OpenBSD.

```
coroutine_backend = get_option('coroutine_backend')
ucontext_probe = '''
  #include <ucontext.h>
  #ifdef __stub_makecontext
  #error Ignoring glibc stub makecontext which will always fail
  #endif
  int main(void) { makecontext(0, 0, 0); return 0; }'''

# On Windows the only valid backend is the Windows specific one.
# For POSIX prefer ucontext, but it's not always possible. The fallback
# is sigcontext.
supported_backends = []
if targetos == 'windows'
  supported_backends += ['windows']
else
  if targetos != 'darwin' and cc.links(ucontext_probe)
    supported_backends += ['ucontext']
  endif
  supported_backends += ['sigaltstack']
endif
```

**Review by vossmjp — APPROVED** [review]
Looks ok to me. Seems better to compile than not for platforms where the only option is the thread version.

**Review by isaevil — APPROVED** [review]
LGTM

**Pull request merged**